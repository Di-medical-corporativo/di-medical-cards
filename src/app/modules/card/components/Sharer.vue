<template>
  <div class="sharer">
    <button
        class="sharer__button"
        data-bs-toggle="modal"
        data-bs-target="#exampleModal"
        :style="sucursalColor('background')">
        <i class="bi bi-share-fill icon"></i>
        Compartir
    </button>
    <div class="modal fade" id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
        <div class="modal-dialog modal-dialog-centered">
            <div class="modal-content">
                <div class="modal-header">
                    <h5 class="modal-title" id="exampleModalLabel">Compartir vcard</h5>
                    <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                </div>
                <div class="modal-body">
                    <div class="sharer__options" :style="sucursalColor('color')">
                        <a :href="`https://www.facebook.com/sharer/sharer.php?u=${url}`" target="_blank"><i class="bi bi-facebook icon"></i></a>
                        <a :href="`https://api.whatsapp.com/send?text=Contacta con ${employeeName} por medio de su Vcard: ${url} | Lo mÃ¡s selecto de las mejores marcas para el mercado hospitalario`" target="_blank"><i class="bi bi-whatsapp icon"></i></a>
                        <i class="bi bi-link icon" @click="copyToClipBoard" data-toggle="tooltip" data-placement="top" title="Tooltip on top"></i>
                    </div>
                </div>
            </div>
        </div>
    </div>

  </div>
</template>

<script>
import { computed } from 'vue'
import { useRoute } from 'vue-router'
import bookMark from '../helpers/bookMarkCard'

export default {
  props: {
    sucursal: {
      type: String
    },
    employeeName: {
      type: String
    }
  },

  setup (props) {
    const route = useRoute()
    const { sucursal, id } = route.params
    return {
      sucursalColor: computed(() => {
        const color = props.sucursal === 'sur' ? '#218d9b' : '#c28400'
        return (typeStyle) => {
          return typeStyle === 'color'
            ? { color: color }
            : { background: color }
        }
      }),
      bookMarkIcon: computed(() => {
        if (!localStorage.getItem('bookmark')) {
          return 'bi bi-bookmark icon'
        }
        return 'bi bi-bookmark-fill icon'
      }),
      saveToFavorites: () => bookMark(),
      url: `https://card.dimedicalcorporativo.mx/${sucursal}/${id}`,
      copyToClipBoard: async () => {
        await navigator.clipboard.writeText(`https://card.dimedicalcorporativo.mx/${sucursal}/${id}`)
        alert('Enlace copiado al portapapeles')
      }
    }
  }
}
</script>
