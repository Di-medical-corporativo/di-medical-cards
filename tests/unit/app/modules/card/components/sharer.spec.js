import { shallowMount } from '@vue/test-utils'
import Sharer from '../../../../../../src/app/modules/card/components/Sharer.vue'

jest.mock('vue-router', () => ({
  useRoute: jest.fn(() => ({
    params: {
      id: 'kevin-aron-tapia-cruz',
      sucursal: 'sur'
    }
  })),
  useRouter: jest.fn(() => ({
    push: jest.fn()
  }))
}))

describe('Sharer component', () => {
  beforeEach(() => {
    let clipboardData = '' // initalizing clipboard data so it can be used in testing
    const mockClipboard = {
      writeText: jest.fn(
        (data) => { clipboardData = data }
      ),
      readText: jest.fn(
        () => { return clipboardData }
      )
    }
    global.navigator.clipboard = mockClipboard
    global.alert = jest.fn()
  })

  test('should match snapshot', () => {
    const wrapper = shallowMount(Sharer, {
      propsData: {
        sucursal: 'sur'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should change background button depending on sucursal param', () => {
    const wrapper = shallowMount(Sharer, {
      propsData: {
        sucursal: 'sur'
      }
    })

    const colorBackground = wrapper.vm.sucursalColor('color')
    expect(colorBackground.color).toBe('#218d9b')
  })

  test('should return object for background when sucursalColor is given background parameter', () => {
    const wrapper = shallowMount(Sharer, {
      propsData: {
        sucursal: 'sur'
      }
    })

    const colorBackground = wrapper.vm.sucursalColor('background')
    expect(colorBackground).toEqual({ background: '#218d9b' })
  })

  test('should copy to clipboard link', () => {
    const wrapper = shallowMount(Sharer, {
      propsData: {
        sucursal: 'sur'
      }
    })
    const copyToClipBoardButton = wrapper.findAll('.icon')[3]
    copyToClipBoardButton.trigger('click')
    expect(global.navigator.clipboard.writeText).toHaveBeenCalled()
  })

  test('should call alert method', async () => {
    const wrapper = shallowMount(Sharer, {
      propsData: {
        sucursal: 'sur'
      }
    })
    const copyToClipBoardButton = wrapper.findAll('.icon')[3]
    await copyToClipBoardButton.trigger('click')
    expect(global.alert).toHaveBeenCalled()
  })
})
