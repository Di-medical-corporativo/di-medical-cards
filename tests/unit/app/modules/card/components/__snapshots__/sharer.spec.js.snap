// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sharer component should match snapshot 1`] = `
<div class="sharer"><button class="sharer__button" data-bs-toggle="modal" data-bs-target="#exampleModal" style="background: rgb(33, 141, 155);"><i class="bi bi-share-fill icon"></i> Compartir </button>
  <div class="modal fade" id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-dialog-centered">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="exampleModalLabel">Compartir vcard</h5><button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
        </div>
        <div class="modal-body">
          <div class="sharer__options" style="color: rgb(33, 141, 155);"><a href="https://www.facebook.com/sharer/sharer.php?u=https://card.dimedicalcorporativo.mx/sur/kevin-aron-tapia-cruz" target="_blank"><i class="bi bi-facebook icon"></i></a><a href="https://api.whatsapp.com/send?text=Contacta con undefined por medio de su Vcard: https://card.dimedicalcorporativo.mx/sur/kevin-aron-tapia-cruz | Lo mÃ¡s selecto de las mejores marcas para el mercado hospitalario" target="_blank"><i class="bi bi-whatsapp icon"></i></a><i class="bi bi-link icon" data-toggle="tooltip" data-placement="top" title="Tooltip on top"></i><a data-bs-dismiss="modal"><i class="bi bi-qr-code-scan icon"></i></a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
